audio/out/ao_avfoundation.m

```diff
@@ -0,0 +1,275 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ao.h"
+#include "audio/format.h"
+#include "audio/out/ao_coreaudio_chmap.h"
+#include "audio/out/ao_coreaudio_utils.h"
+#include "common/common.h"
+#include "common/msg.h"
+#include "internal.h"
+#include "osdep/timer.h"
+#include "ta/ta_talloc.h"
+
+#import <AVFoundation/AVFoundation.h>
+#include <Foundation/Foundation.h>
+#import <CoreAudioTypes/CoreAudioTypes.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <CoreMedia/CoreMedia.h>
+
+struct priv {
+    AVSampleBufferAudioRenderer *renderer;
+    AVSampleBufferRenderSynchronizer *synchronizer;
+    dispatch_queue_t queue;
+    CMAudioFormatDescriptionRef format_description;
+    int64_t end_time_mp;
+    int64_t end_time_av;
+};
+
+static int64_t CMTimeGetMicroseconds(CMTime time)
+{
+    time = CMTimeConvertScale(time, 1000000, kCMTimeRoundingMethod_Default);
+    return time.value;
+}
+
+static void feed(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->end_time_mp == -1) {
+        p->end_time_mp = mp_time_us();
+    }
+    if (p->end_time_av == -1) {
+        p->end_time_av = CMTimeGetMicroseconds([p->synchronizer currentTime]);
+    }
+
+    int sample_count = ao->samplerate / 10; // 1/10 samplerate may be a good number
+    int buffer_size = sample_count * ao->sstride;
+    CMBlockBufferRef block_buffer = NULL;
+    if (CMBlockBufferCreateWithMemoryBlock(
+        kCFAllocatorDefault,
+        NULL,
+        buffer_size,
+        NULL,
+        NULL,
+        0,
+        buffer_size,
+        kCMBlockBufferAssureMemoryNowFlag,
+        &block_buffer
+    ) != noErr) {
+        MP_FATAL(ao, "failed to create CMBlockBuffer\n");
+        goto error;
+    }
+    void* data[1];
+    if (CMBlockBufferGetDataPointer(block_buffer, 0, NULL, NULL, (char**)data) != noErr) {
+        MP_FATAL(ao, "CMBlockBufferGetDataPointer failed\n");
+        goto error;
+    }
+
+    ao_read_data(ao, data, sample_count, p->end_time_mp);
```

Comment:
Many ao, including coreaudio and audiounit, do not check the return value of `ao_read_data` as well. I don't think filling the buffer with silence when underruns happen is unacceptable. And the `device_buffer` of this ao is set to be large enough to avoid underruns in normal situation.

If we feed the renderer with actual number of samples, when underruns happen, the renderer may get fewer samples than what it expects. In this case, the requesting queue will be constantly waken by the renderer, which causes spinning.